lib_test/test_httpaf.ml

```diff
@@ -154,9 +198,121 @@ module Server_connection = struct
   ;;
 
+  let test_synchronous_error () =
+    let request_string = "/ GET HTTP/1.1\r\n\r\n" in
+    let writer_woken_up = ref false in
+    let t = create ~error_handler synchronous_raise in
+    Alcotest.(check (of_pp Write_operation.pp_hum)) "Writer is in a yield state"
+      `Yield (next_write_operation t);
+    yield_writer t (fun () -> writer_woken_up := true);
+    let c = read_string t request_string in
+    Alcotest.(check int) "read consumes all input"
+      (String.length request_string) c;
+    Alcotest.(check (of_pp Read_operation.pp_hum)) "Error shuts down the reader"
+      `Close (next_read_operation t);
+    Alcotest.(check bool) "Writer woken up"
+      true !writer_woken_up;
+    Alcotest.(check (option string)) "Error response written"
+      (Some "HTTP/1.1 500 Internal Server Error\r\n\r\ngot an error")
+      (next_write_operation t |> Write_operation.to_write_as_string)
+  ;;
+
+  let test_synchronous_error_asynchronous_handling () =
+    let request_string = "/ GET HTTP/1.1\r\n\r\n" in
+    let writer_woken_up = ref false in
+    let continue = ref (fun () -> ()) in
+    let error_handler ?request error start_response =
+      continue := (fun () ->
+        error_handler ?request error start_response)
+    in
+    let t = create ~error_handler synchronous_raise in
+    Alcotest.(check (of_pp Write_operation.pp_hum)) "Writer is in a yield state"
+      `Yield (next_write_operation t);
+    yield_writer t (fun () -> writer_woken_up := true);
+    let c = read_string t request_string in
+    Alcotest.(check int) "read consumes all input"
+      (String.length request_string) c;
+    Alcotest.(check (of_pp Write_operation.pp_hum)) "Writer is in a yield state"
+      `Yield (next_write_operation t);
+    !continue ();
+    Alcotest.(check (of_pp Read_operation.pp_hum)) "Error shuts down the reader"
+      `Close (next_read_operation t);
+    Alcotest.(check bool) "Writer woken up"
+      true !writer_woken_up;
+    Alcotest.(check (option string)) "Error response written"
+      (Some "HTTP/1.1 500 Internal Server Error\r\n\r\ngot an error")
+      (next_write_operation t |> Write_operation.to_write_as_string)
+  ;;
+
+
+  let test_asynchronous_error () =
+    let continue = ref (fun () -> ()) in
+    let asynchronous_raise reqd =
+      continue := (fun () -> synchronous_raise reqd)
+    in
+    let request_string = "/ GET HTTP/1.1\r\n\r\n" in
+    let writer_woken_up = ref false in
+    let t = create ~error_handler asynchronous_raise in
+    Alcotest.(check (of_pp Write_operation.pp_hum)) "Writer is in a yield state"
+      `Yield (next_write_operation t);
+    yield_writer t (fun () -> writer_woken_up := true);
+    let c = read_string t request_string in
+    Alcotest.(check int) "read consumes all input"
+      (String.length request_string) c;
+    Alcotest.(check (of_pp Write_operation.pp_hum)) "Writer is in a yield state"
+      `Yield (next_write_operation t);
+    !continue ();
```

Comment:
Before we continue, should we assert what the read state is?

Reply:
Done, in both async raising cases.